---
title: "10xgenomics"
output: html_document
---

## Prepare session

```{r setup, include=FALSE}
require(knitr)
knitr::opts_chunk$set(echo = TRUE)
path <- '~/Desktop/10xgenomics'
dir.create(path = path, recursive = T)
knitr::opts_knit$set(root.dir = path, collapse = T, comment = '#>')

# Load libraries
library(Seurat)
library(sctransform)
library(dplyr)
library(ggplot2)
#library(SeuratWrappers)

# Define custom themes for ggplot
theme_custom <- theme(axis.text.x = element_text(size = 12.8, color = 'black'),
                      axis.text.y = element_text(size = 12.8, color = 'black'), 
                      axis.title.x = element_text(size = 16, margin = margin(6,0,0,0)), 
                      axis.title.y = element_text(size = 16, margin = margin(0,8,0,0)),
                      plot.title = element_text(size = 16, face = 'plain', hjust = 0.5, margin = margin(5,0,10,0)),
                      plot.subtitle = element_text(size = 12, hjust = 0.5, margin = margin(0,0,10,0)), 
                      axis.ticks.x = element_line(size = 0.4, colour = 'black'), 
                      axis.ticks.y = element_line(size = 0.4, colour = 'black'),
                      axis.line = element_blank(), 
                      panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(),
                      panel.grid.minor.x = element_blank(), panel.grid.major.x = element_blank(),
                      panel.border = element_rect(colour = 'black', size = 0.8, fill = NA),
                      panel.background = element_rect(colour = NA, fill = NA), 
                      strip.text.x = element_text(size = 16, face = 'plain', margin = margin(0,0,5,0) ), 
                      strip.text.y = element_text(size = 16, face = 'plain', margin = margin(0,0,0,5) ), 
                      strip.background = element_blank(), 
                      legend.key = element_blank(), legend.key.size = unit(0.5,'cm'),
                      legend.text = element_text(size = 12.8), 
                      legend.title = element_text(size = 12.8),
                      plot.background = element_rect(colour = NA, fill = NA), aspect.ratio = 1)
theme_custom2 <- theme(axis.text.x = element_blank(), 
                       axis.text.y = element_blank(), 
                       axis.title.x = element_text(size = 16, margin = margin(4,0,0,0)), 
                       axis.title.y = element_text(size = 16, margin = margin(0,4,0,0)),
                       plot.title = element_text(size = 16, face = 'plain', hjust = 0.5, margin = margin(5,0,10,0)),
                       plot.subtitle = element_text(size = 12, hjust = 0.5, margin = margin(0,0,10,0)), 
                       axis.ticks.x = element_blank(), 
                       axis.ticks.y = element_blank(), 
                       axis.line = element_blank(), 
                       panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(),
                       panel.grid.minor.x = element_blank(), panel.grid.major.x = element_blank(),
                       panel.border = element_rect(colour = 'grey80', size = 0.8, fill = NA),
                       panel.background = element_rect(colour = NA, fill = NA),
                       strip.text.x = element_text(size = 16, face = 'plain', margin = margin(0,0,5,0) ), 
                       strip.text.y = element_text(size = 16, face = 'plain', margin = margin(0,0,0,5) ), 
                       strip.background = element_blank(), 
                       legend.key = element_blank(), legend.key.size = unit(0.5,'cm'),
                       legend.text = element_text(size = 12.8), 
                       legend.title = element_text(size = 12.8),
                       plot.background = element_rect(colour = NA, fill = NA), aspect.ratio = 1)

# Function to save plots with same size
grid <- function(data, layout=rbind(c(1,1,1,1,2,2,3))) {
  gridExtra::grid.arrange(grobs = list(data + guides(color = 'none', fill = 'none', size = 'none'), 
                                       cowplot::get_legend(data)), ncol = 2, layout_matrix = layout, 
                          top = grid::textGrob('', gp = grid::gpar(fontsize = 16, font = 2)))
}

# Function to remove clip from strip labels
rmclip <- function(plot,gg=F) {
  pg <- ggplotGrob(plot)
  for(i in which(grepl("strip-t", pg$layout$name))) {
    pg$grobs[[i]]$layout$clip <- "off"
  }
  if(gg) return(pg) else return(grid::grid.draw(pg))
}
```

-------------------------------------------------------------------------------------------------------------------

## All cells

- Merge the 4 replicates and both conditions together

### Clustering with Seurat

```{r procdata, collapse=T, echo=F, warning=F, message=F, include=F}
# Download pre-processed data from GEO SuperSeries and place at '~/Desktop/10xgenomics/'
download.file(url='https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE200488&format=file', 
              destfile=paste0(path, '/GSE200488_RAW.tar'))
untar(tarfile = paste0(path, '/GSE200488_RAW.tar'), exdir = paste0(path) )
sapply(list.files(path, full.names=T)[grepl('GSM', list.files(path))], function(x) untar(tarfile=x, exdir=paste0(path) ))

# Load UMI counts from 10x data
dirs <- list.dirs(paste0(path,'/cellranger'))
dirs <- dirs[grep('filtered_feature_bc_matrix', dirs)]
names.dirs <- stringr::str_split(dirs, pattern = '/', n = Inf, simplify = T)
names(dirs) <- sub('_filtered_feature_bc_matrix', '', sub('-Spleen', '', names.dirs[,ncol(names.dirs)]) )
names(dirs) <- sub('5', '4', names(dirs))
bd <- lapply(dirs, function(x) Read10X(data.dir = x) )

# Initialize the Seurat object with the raw (non-normalized data).
bd <- lapply(setNames(names(bd), nm=names(bd)), function(x) 
  AddMetaData(CreateSeuratObject(counts = bd[[x]], min.cells = 3, min.features = 0, project = 'ot1_lefl_vs_cmc'), 
              metadata = x, col.name = 'sample') )
# Rename cells
bd <- Map(x=names(bd), function(x) RenameCells(bd[[x]], new.names = paste0(bd[[x]]$sample, '_', colnames(bd[[x]])) ))

# Define colors
col <- list(sp=setNames(c('#D78184', '#7BAFDE', '#96C793', '#b888ae', '#b00409', '#1965B0', '#2d9027', '#6c245b', '#FF7F00', '#ffbf7f', '#b9206f', '#E78AC3', '#55A1B1', '#8DD3C7', '#A6761D', '#E6AB02', '#7570B3', '#BEAED4', '#666666', '#999999', '#997575', '#d4b7b7', '#737300', '#aeae5c'), nm=unique(gsub('_',' ',names(dirs))) ), 
            cl=setNames(c('#b00409', '#D78184', '#1965B0', '#7BAFDE', '#2d9027', '#96C793', '#6c245b'), nm=c('3','5','2','4a','4b','4c','1') ), 
            gp=setNames(c('grey60','royalblue3'), nm=unique(gsub('_',' ',sub('\\d','',names(dirs)))) ), 
            ct=c(OT1='grey80', other='grey80', 'B'='#96C793','B/M'='#96C793','Cyc'='#7570B3','rRNA'='#E78AC3','M'='#E6AB02', 'mtRNA'='#7BAFDE'), 
            hm=colorRampPalette(c('#133d67','#2166AC','#90b2d5','#d2e0ee','grey90','#efd0d4','#d88b95','#B2182B','#590c15'))(41), 
            hm2=colorRampPalette(c('grey90','#efd0d4','#d88b95','#B2182B','#590c15'))(41) )

# Calculate the percentage of mitochondrial features here and store it in object metadata as `percent.mito`
bd <- lapply(bd, FUN = PercentageFeatureSet, pattern = "^mt-", col.name = "percent.mito")

# Keep cells that have unique feature counts less than 3 SDs above the mean, less than 10% of mitochondrial genes 
# and more than 500 detected genes
bd <- lapply(bd, FUN = subset, subset = nFeature_RNA > 500 & nCount_RNA < summary(nCount_RNA)[4]+sd(nCount_RNA)*3 & percent.mito < 10)

# Normalize all samples together
bd <- Map(x='merged', function(x) {
  bdI <- merge(bd[[1]], y = bd[-1])
  bdI <- SCTransform(bdI, method = 'glmGamPoi', variable.features.n = 1000, assay = 'RNA', new.assay.name = 'SCTmerged', return.only.var.genes = F)
  bdI <- PercentageFeatureSet(bdI, pattern = "^mt-", col.name = "percent.mito_SCTmerged")
  
  # Add metadata
  bdI@meta.data <- cbind(bdI@meta.data, data.frame(condition = stringr::str_split(bdI$sample, pattern='\\d', n=Inf, simplify=T)[,1], replicate = sub('CMC|Lefl', '', bdI$sample)))
  bdI$condition <- factor(bdI$condition, levels=sort(unique(sub('[[:digit:]]','',names(dirs)))) )
  bdI$sample <- factor(bdI$sample, levels=sort(unique(names(dirs))) )
  
  # Perform linear dimensional reduction
  bdI <- RunPCA(bdI, features = VariableFeatures(bdI), verbose = T, assay = 'SCTmerged', reduction.name = 'mpca', reduction.key = 'mPC_')
  
  # Compute KNN graphs
  bdI <- FindNeighbors(bdI, reduction = 'mpca', k.param = 20, dims = 1:20, do.plot = F, assay = 'SCTmerged')
  return(bdI)
})

# Visualization and clustering
bd <- Map(x=names(bd), function(x) {
  # Cluster cells
  bdI <- FindClusters(bd[[x]], resolution = 0.33, algorithm = 1, n.start = 10, n.iter = 10, random.seed = 0)
  bdI$seurat_clusters <- factor(plyr::mapvalues(as.numeric(bdI$seurat_clusters), from=sort(unique(as.numeric(bdI$seurat_clusters))), to=1:length(unique(bdI$seurat_clusters)) ))
  colnames(bdI@meta.data)[colnames(bdI@meta.data) %in% 'seurat_clusters'] <- 'cluster'
  Idents(bdI) <- 'cluster'
  
  # Run Non-linear dimensional reduction (tSNE and UMAP)
  bdI <- RunTSNE(bdI, reduction = 'mpca', dims = 1:20, assay = 'SCTmerged', reduction.name = 'mtsne', reduction.key = 'mtSNE_', seed.use = 1, tsne.method = 'Rtsne', dim.embed = 2, perplexity = 30)
  bdI <- RunUMAP(bdI, reduction = 'mpca', dims = 1:20, assay = 'SCTmerged', reduction.name = 'mumap', reduction.key = 'mUMAP_', seed.use = 3, spread = 1, min.dist = 0.3)
  #DimPlot(bdI, reduction = 'mumap', dims = c(1,2), group.by = 'sample', split.by = 'orig.ident', label = F, label.size = 8, pt.size = 0.3, cols = col$sp) + theme_bw() + theme_custom2 #+ NoLegend()
  return(bdI)
})
#trqwe::mcsaveRDS(bd, file = paste0(path, '/seurat.rds'), mc.cores = min(parallel::detectCores(), 14))
#bd <- trqwe::mcreadRDS(paste0(path, '/seurat.rds'), mc.cores = min(parallel::detectCores(), 14))

# Plot clusters
s <- 'merged'
umm.cl <- DimPlot(bd[[s]], reduction = 'mumap', dims = c(1,2), group.by = 'cluster', label = T, label.box = T, label.size = 3, pt.size = 0.1, cols = as.vector(col$sp)) + ggtitle('OT1 cells', subtitle = paste0(gsub('_',' ',s), ' (non-integrated)')) + theme_bw() + theme_custom2 + scale_x_continuous('UMAP 1',breaks=scales::pretty_breaks()) + scale_y_continuous('UMAP 2',breaks=scales::pretty_breaks())
```

### Cluster annotation

```{r procdataanno, collapse=T, echo=F, warning=F, message=F, include=F}
# Function to draw dotplot
dotplot <- function(object=bd, sel=genes$ct, cl=F, scale=T, sample='OT1 CD8+ T cells', subtitle='Marker expression by cluster',smax=100) {
  sel <- sel[sel %in% rownames(object)]
  dt <- DotPlot(object, assay = 'SCTmerged', features = sel, cluster.idents = cl, scale.max=smax)$data
  if(scale) l <- list(a=-1.5,b=1.5) else l <- list(a=0,b=6)
  ggplot(transform(dt, celltype= plyr::mapvalues(dt$id, from=sort(unique(as.numeric(dt$id))), to=ctmap[[s]])), 
         aes(id, features.plot, color=if(scale) avg.exp.scaled else avg.exp, size=pct.exp)) + 
    geom_point() + scale_x_discrete('') + scale_y_discrete('') + #coord_flip() + 
    scale_color_gradientn('Average expression', colours = col$hm, oob = scales::squish, 
                          limits = c(l$a,l$b), breaks = seq(l$a,l$b, length.out = 3)) + 
    scale_size_continuous('Percent expressed', limits = c(0,smax), breaks = seq(0,smax,smax/4), range = c(0,6)) + 
    ggtitle(gsub('_',' ',sample), subtitle = subtitle) + theme_custom + 
    facet_grid(.~factor(celltype, levels = levels(object$celltype)), scales = 'free', space = 'free') + 
    theme(aspect.ratio = length(sel) / length(unique(dt$id)), 
          strip.text = element_text(size = 12.8, face = 'plain', angle=90, vjust=0.5, hjust=0, margin = margin(5,0,5,0)), 
          axis.text.x = element_text(size = 12.8, color = 'black', angle = 0, vjust = 1, hjust = 0.5))
}

# Identify celltypes
ctmap <- list(merged=c('1'='other','2'='Cyc','3'='other','4'='rRNA','5'='Cyc','6'='other','7'='mtRNA','8'='B/M') )
bd[[s]]$celltype <- plyr::mapvalues(bd[[s]]$cluster, from=names(ctmap[[s]]), to=ctmap[[s]])
bd[[s]]$celltype <- factor(bd[[s]]$celltype, levels = sort(unique(as.character(bd[[s]]$celltype)))[c(4,2,5,3,1)])

# Plot celltypes over UMAP
um.ct <- DimPlot(bd[[s]], reduction = 'mumap', dims = c(1,2), group.by = 'celltype', label = F, label.size = 5, pt.size = 0.1, cols = (col$ct)) + ggtitle('OT1 cells', subtitle = paste0(gsub('_',' ',s), ' (integrated)')) + xlab('UMAP 1') + ylab('UMAP 2') + theme_bw() + theme_custom2 + scale_x_continuous(breaks=scales::pretty_breaks()) + scale_y_continuous(breaks=scales::pretty_breaks())

# Plot celltype markers
genes <- list(ct=c('AY036118','Gm42418','Cdk1','Mcm2','Pclaf','H2afz','Birc5','Mki67','mt-Atp6','mt-Co1','mt-Co2','mt-Co3','mt-Nd3','mt-Nd4','Cd14','Lyz2','Fcgr3','Ms4a7','Fcer1g','Cst3','H2-Aa','Ly6d','Ms4a1','Cd19') )
dp.ct <- dotplot(object = bd[[s]], sel = genes$ct, sample=s, subtitle = 'Cell type marker expression by cluster')

# Export cell annotation
write.csv(bd[[s]]@meta.data, file = paste0(path, '/cell_annotation.csv'), row.names = T)
```

-------------------------------------------------------------------------------------------------------------------

## Filtered data

- Merge both conditions together and integrate the 4 replicates
- Filter out clusters of contaminating cells identified on previous analysis

### Clustering with Seurat

```{r filterdata, collapse=T, echo=F, warning=F, message=F, include=F}
# Load UMI counts from 10x data
bd <- lapply(dirs, function(x) Read10X(data.dir = x) )

# Initialize the Seurat object with the raw (non-normalized data).
bd <- lapply(setNames(names(bd), nm=names(bd)), function(x) 
    AddMetaData(CreateSeuratObject(counts = bd[[x]], min.cells = 3, min.features = 0, project = 'ot1_lefl_vs_cmc'), 
                metadata = x, col.name = 'sample') )
# Rename cells
bd <- Map(x=names(bd), function(x) RenameCells(bd[[x]], new.names = paste0(bd[[x]]$sample, '_', colnames(bd[[x]])) ))

# Calculate the percentage of mitochondrial features here and store it in object metadata as `percent.mito`
bd <- lapply(bd, FUN = PercentageFeatureSet, pattern = "^mt-", col.name = "percent.mito")

# Keep cells that have unique feature counts less than 3 SDs above the mean, less than 10% of mitochondrial genes 
# and more than 2000 detected genes
bd <- lapply(bd, FUN = subset, subset = nFeature_RNA > 2000 & nCount_RNA < summary(nCount_RNA)[4]+sd(nCount_RNA)*3 & percent.mito < 10)

# Remove contaminating cells (based on prior analysis over merged data)
ann <- read.csv(file = paste0(path, '/cell_annotation.csv'), row.names = 1)
ann <- ann[!ann$celltype %in% c('mtRNA','B/M'),]
bd <- Map(x=names(bd), function(x) subset(
 AddMetaData(bd[[x]], metadata = colnames(bd[[x]]), col.name = 'cells'), subset = cells %in% rownames(ann)) )

# Group by replicate number
bd <- Map(x=unique(sub('CMC|Lefl','',names(bd))), function(x) bd[sub('CMC|Lefl','',names(bd)) %in% x])

# Merge conditions by replicate number
bd <- Map(x=names(bd), function(x) merge(bd[[x]][[1]], y = bd[[x]][-1]) )

# Normalize merged conditons together
bd <- lapply(bd, FUN = SCTransform, method = 'glmGamPoi', variable.features.n = 1000)

# Identify anchor points and integrate datasets
bd <- Map(x='integrated', function(x) {
 int.features <- SelectIntegrationFeatures(object.list = bd, nfeatures = 1000)
 bdI <- PrepSCTIntegration(object.list = bd, anchor.features = int.features )
 bdI <- FindIntegrationAnchors(object.list = bdI, normalization.method = "SCT", k.anchor = 5, anchor.features = int.features)
 bdI <- IntegrateData(anchorset = bdI, dims = 1:20, normalization.method = "SCT", features.to.integrate = unique(unlist(sapply(bdI@object.list, rownames))) )
 
 # Add metadata
 bdI@meta.data <- cbind(bdI@meta.data, data.frame(condition = stringr::str_split(bdI$sample, pattern='\\d', n=Inf, simplify=T)[,1], 
                                                  replicate = sub('CMC|Lefl', '', bdI$sample)))
 bdI$condition <- factor(bdI$condition, levels=sort(unique(sub('[[:digit:]]','',names(dirs)))) )
 bdI$sample <- factor(bdI$sample, levels=sort(unique(names(dirs))) )
 
 # Perform linear dimensional reduction
 bdI <- RunPCA(bdI, features = VariableFeatures(bdI), verbose = F, assay = 'integrated')
 
 # Compute KNN graphs
 bdI <- FindNeighbors(bdI, reduction = 'pca', k.param = 30, dims = 1:20, do.plot = F, assay = 'integrated')
 return(bdI)
})

# Visualization and clustering
bd <- Map(x=names(bd), function(x) {
 bdI <- FindClusters(bd[[x]], resolution = 0.33, algorithm = 1, n.start = 10, n.iter = 10, random.seed = 0)
 bdI$seurat_clusters <- factor(plyr::mapvalues(as.numeric(bdI$seurat_clusters), from=sort(unique(as.numeric(bdI$seurat_clusters))), 
                                               to=1:length(unique(bdI$seurat_clusters)) ))
 colnames(bdI@meta.data)[colnames(bdI@meta.data) %in% 'seurat_clusters'] <- 'cluster'
 
 # Change cluster numbers
 bdI$cluster <- factor(plyr::mapvalues(as.numeric(bdI$cluster), from=sort(unique(as.numeric(bdI$cluster)))[c(4,1,3,2,6,5,7)], #[c(3,0,2,1,5,4,6)], 
                                       to=1:length(unique(bdI$cluster)) ))
 
 # Run Non-linear dimensional reduction (tSNE and UMAP)
 bdI <- RunTSNE(bdI, reduction = 'pca', dims = 1:20, seed.use = 2, tsne.method = 'Rtsne', dim.embed = 2, perplexity = 30, assay = 'integrated')
 bdI <- RunUMAP(bdI, reduction = 'pca', dims = 1:20, seed.use = 0, spread = 1, min.dist = 0.1, reduction.name = 'umap', assay = 'integrated')
 #DimPlot(bdI, reduction = 'umap', dims = c(1,2), group.by = 'cluster', split.by = 'condition', label = F, label.size = 8, pt.size = 0.3, cols = as.vector(col$cl)) + theme_bw() + theme_custom2 #+ NoLegend()
 return(bdI)
})

# Normalize all datasets together
bd <- Map(x=names(bd), function(x) {
 bdI <- SCTransform(bd[[x]], method = 'glmGamPoi', variable.features.n = 1000, assay = 'RNA', new.assay.name = 'SCTmerged', return.only.var.genes = F)
 Idents(bdI) <- 'cluster'
 bdI <- PercentageFeatureSet(bdI, pattern = "^mt-", col.name = "percent.mito_SCTmerged")
 
 # Perform linear dimensional reduction
 bdI <- RunPCA(bdI, features = VariableFeatures(bdI), verbose = T, assay = 'SCTmerged', reduction.name = 'mpca', reduction.key = 'mPC_')
 
 # Run Non-linear dimensional reduction (tSNE and UMAP)
 bdI <- RunTSNE(bdI, reduction = 'mpca', dims = 1:20, assay = 'SCTmerged', reduction.name = 'mtsne', reduction.key = 'mtSNE_', seed.use = 1, tsne.method = 'Rtsne', dim.embed = 2, perplexity = 30)
 bdI <- RunUMAP(bdI, reduction = 'mpca', dims = 1:20, assay = 'SCTmerged', reduction.name = 'mumap', reduction.key = 'mUMAP_', seed.use = 0, spread = 1, min.dist = 0.1)
 #DimPlot(bdI, reduction = 'mumap', dims = c(1,2), group.by = 'cluster', split.by = 'condition', label = F, label.size = 8, pt.size = 0.3, cols = as.vector(col$cl)) + theme_bw() + theme_custom2 #+ NoLegend()
 return(bdI)
})

# Update cluster labels
s<-'integrated'
bd[[s]]$cluster_old <- bd[[s]]$cluster
bd[[s]]$cluster <- plyr::mapvalues(bd[[s]]$cluster, from = levels(bd[[s]]$cluster), to = c('3','5','2','4a','4b','4c','1') )
bd[[s]]$cluster <- factor(bd[[s]]$cluster, levels = sort(levels(bd[[s]]$cluster)) )
Idents(bd[[s]]) <- 'cluster'
#trqwe::mcsaveRDS(bd, file = paste0(path, '/seurat_filtered.rds'), mc.cores = min(parallel::detectCores(), 14))
#bd <- trqwe::mcreadRDS(paste0(path, '/seurat_filtered.rds'), mc.cores = min(parallel::detectCores(), 14))

# Calculate average cluster expression
avg <- list(norm=AverageExpression(object=AddMetaData(object=bd[[s]], metadata=paste0('cl', bd[[s]]$cluster), col.name='cl'), assay='SCTmerged', return.seurat=F, group.by='cl', slot='data')[[1]], score=AverageExpression(object=AddMetaData(object=bd[[s]], metadata=paste0('cl', bd[[s]]$cluster), col.name='cl'), assay='SCTmerged', return.seurat=F, group.by='cl', slot='scale.data')[[1]] )
avg <- Map(x=names(avg), function(x) transform(as.data.frame(avg[[x]]), gene=rownames(avg[[x]]), row.names=1:nrow(avg[[x]]))[,c((length(unique(bd[[s]]$cluster))+1),1:(length(unique(bd[[s]]$cluster))))] %>% arrange(gene) )

# Export average cluster expression
openxlsx::write.xlsx(avg, firstRow=T, firstCol=T, asTable=T, withFilter=T, 
                     file = paste0(path, '/cluster_avg_expression.xlsx'), startRow = 1, #sheet = 1, 
                     colNames = T, rowNames = F, rows = NULL, cols = NULL, namedRegion = NULL, keepNA = T, na.string='N/A')

# Export metadata
md <- list(OT1_Lefl_vs_CMC=transform(as.data.frame(Reduce(merge, lapply(list(bd[[s]]@meta.data[,], Seurat::Embeddings(bd[[s]], reduction = "umap"), t(as.matrix(Seurat::GetAssayData(bd[[s]], assay = 'SCT', slot = 'data')))), data.table::data.table, keep.rownames = TRUE, key = 'rn') )), row.names=rn) %>% dplyr::rename(cell=rn) )
lapply(names(md), function(x) 
  data.table::fwrite(md[[x]], file = paste0(path, '/GSE200488_OT1_Lefl_vs_CMC_norm.txt.gz'), sep = '\t', row.names = F, col.names = T, compress = 'gzip', nThread = min(parallel::detectCores(), 14) ) )

# Plot data
um.cls <- DimPlot(bd[[s]], reduction = 'umap', dims = c(1,2), group.by = 'cluster', split.by = 'condition', label = T, label.size = 5, pt.size = 0.1, cols = col$cl, ncol=4) + ggtitle('OT1 cells', subtitle = paste0(gsub('_',' ',s), ' (integrated)')) + xlab('UMAP 1') + ylab('UMAP 2') + theme_bw() + theme_custom2 + scale_x_continuous(breaks=scales::pretty_breaks()) + scale_y_continuous(breaks=scales::pretty_breaks()) #ExtData10A
um.clr <- DimPlot(bd[[s]], reduction = 'umap', dims = c(1,2), group.by = 'cluster', split.by = 'sample', label = T, label.size = 5, pt.size = 0.1, cols = col$cl, ncol=4) + ggtitle('OT1 cells', subtitle = paste0(gsub('_',' ',s), ' (integrated)')) + xlab('UMAP 1') + ylab('UMAP 2') + theme_bw() + theme_custom2 + scale_x_continuous(breaks=scales::pretty_breaks()) + scale_y_continuous(breaks=scales::pretty_breaks()) #ExtData10B
```

### Cell distribution

```{r filterdatadist, collapse=T, echo=F, warning=F, message=F, include=F}
# Calculate frequencies for groups in each cluster
freq.spcl <- transform(as.data.frame(table(group = droplevels(bd[[s]]$sample), cluster = bd[[s]]$cluster)) )
freq.spcl <- freq.spcl[order(freq.spcl$cluster),]

# Calculate percentage from total cells accounting for disparity of total cell number
tot.sp <- plyr::ddply(freq.spcl, 'group', summarise, sum = sum(Freq))
for(i in 1:nrow(freq.spcl)) freq.spcl$percent[i] <- 100*freq.spcl$Freq[i] / 
  tot.sp$sum[tot.sp$group == freq.spcl$group[i]]

# Calculate statistics
freq.spcl$condition <- factor(stringr::str_split(freq.spcl$group, pattern = '\\d', simplify = T)[,1], levels=levels(bd[[s]]$condition))
freq.spcl$replicate <- factor(sub('CMC|Lefl', '', freq.spcl$group))
pval.spcl <- data.frame(freq.spcl %>% group_by(cluster) %>% summarise(max=max(percent)), 
                             padj=round(p.adjust(sapply(levels(bd[[s]]$cluster), function(x) 
                               pairwise.t.test(freq.spcl %>% group_by(group=paste0(cluster,'_',group)) %>% filter(cluster %in% x) %>% pull(percent), 
                                               freq.spcl %>% group_by(group=paste0(cluster,'_',group)) %>% filter(cluster %in% x) %>% pull(condition), 
                                               p.adjust.method = 'none', paired = F)$p.value), method = 'BH'), digits = 3) )
pval.spcl <- pval.spcl[order(pval.spcl$cluster),]
pval.spcl$y <- pval.spcl$max+4
pval.spcl$x <- as.numeric(pval.spcl$cluster)
pval.spcl$xe <- as.numeric(pval.spcl$cluster)

# Save p-values
pval.spcl.save <- data.frame(freq.spcl %>% group_by(cluster) %>% summarise(max=max(percent)), 
                                  p=sapply(levels(bd[[s]]$cluster), function(x) 
                                    pairwise.t.test(freq.spcl %>% group_by(group=paste0(cluster,'_',group)) %>% filter(cluster %in% x) %>% pull(percent), 
                                                    freq.spcl %>% group_by(group=paste0(cluster,'_',group)) %>% filter(cluster %in% x) %>% pull(condition), 
                                                    p.adjust.method = 'none', paired = F)$p.value), 
                                  padj=(p.adjust(sapply(levels(bd[[s]]$cluster), function(x) 
                                    pairwise.t.test(freq.spcl %>% group_by(group=paste0(cluster,'_',group)) %>% filter(cluster %in% x) %>% pull(percent), 
                                                    freq.spcl %>% group_by(group=paste0(cluster,'_',group)) %>% filter(cluster %in% x) %>% pull(condition), 
                                                    p.adjust.method = 'none', paired = F)$p.value), method = 'BH') ) )[,-2]
openxlsx::write.xlsx(pval.spcl.save, firstRow=T, firstCol=T, asTable=T, withFilter=T, 
                     file = paste0(path, '/distribution_fdr.xlsx'), startRow = 1, #sheet = 1, 
                     colNames = T, rowNames = F, rows = NULL, cols = NULL, namedRegion = NULL, keepNA = T, na.string='N/A')

# Save distribution
freq.spcl.save <- transform(freq.spcl, sample=group, absolute=Freq, percent=percent, group=NULL, Freq=NULL)[,c(5,1,3,4,6,2)]
openxlsx::write.xlsx(list('distribution' = freq.spcl.save),
                     firstRow=T, firstCol=T, asTable=T, withFilter=T,
                     file = paste0(path, '/distribution.xlsx'), startRow = 1, colNames = T,
                     rowNames = F, rows = NULL, cols = NULL, namedRegion = NULL, keepNA = T, na.string='N/A')

# Plot distribution
dist.spclds <- ggplot(freq.spcl) + 
  geom_point(aes(x = condition, y = percent, color = condition, fill = condition, shape = replicate), 
             position = position_jitterdodge(jitter.width = 0.1, jitter.height = 0, dodge.width = 0.4, seed = 2), 
             size = 1.6, stat = 'identity', alpha = 0.9, stroke=0.5) + 
  geom_segment(data = pval.spcl, aes(x=1.5+0.4,xend=1.5-0.4,y=y,yend=y)) + 
  geom_text(data = pval.spcl, aes(x = 1.5, y = max+8, label = padj), size = 3, angle = 0, hjust = 0.5) + 
  scale_x_discrete('') + 
  scale_y_continuous('% of total cells', limits = c(0,60), breaks = seq(0,60,10), expand = c(0.03,0)) + 
  scale_color_manual('Condition', values = col$gp) + 
  scale_fill_manual('Condition', values = col$gp, guide = guide_legend(override.aes = list(size = 4, alpha = 0.9))) + 
  scale_shape_manual('Replicate', values = c('1'=21,'2'=22,'3'=23,'4'=24), 
                     guide = guide_legend(override.aes = list(size = 4, alpha = 0.9, color='grey60', fill='grey60'))) +
  facet_grid(.~cluster, scales = 'free', space = 'free') + 
  ggtitle('OT1 cells', subtitle = 'Relative cluster distribution per sample') + theme_custom + 
  theme(aspect.ratio = NULL,axis.text.x = element_text(size = 14, color = 'black', angle = 90, vjust=0.5, hjust = 1)) #ExtData10C
```

### Expression of target genes

```{r filterdataexp, collapse=T, echo=F, warning=F, message=F, include=F}
# Function to draw dotplot
dotplot2 <- function(object=bd, sel=genes$sel1, cl=F, scale=T, b=15, gr=NULL, sample='OT1 cells', subtitle='Expression by cluster',smax=100) {
  sel <- sel[sel %in% rownames(object)]
  dt <- DotPlot(object, assay = 'SCTmerged', features = sel, cluster.idents = cl, scale.max=smax, group.by = gr)$data
  if(!is.null(gr)) dt <- cbind(dt, reshape2::colsplit(dt$id, pattern = '_', names = c('condition','cluster')))
  if(scale) l <- list(a=-1.5,b=1.5) else l <- list(a=0,b=b)
  p <- ggplot(dt, 
              aes(if(!is.null(gr)) factor(cluster, levels=rev(unique(cluster))) else factor(id, levels=rev(unique(id))), features.plot, color=if(scale) avg.exp.scaled else avg.exp, size=pct.exp)) + 
    geom_point() + scale_x_discrete('cluster') + scale_y_discrete('') + coord_flip() + 
    scale_color_gradientn(ifelse(scale,'z-score','Avg. exp.'), colours = if(scale) col$hm else col$hm2, oob = scales::squish, 
                          limits = c(l$a,l$b), breaks = seq(l$a,l$b, length.out = 3)) + 
    scale_size_continuous('Pct. exp', limits = c(0,smax), breaks = seq(0,smax,smax/4), range = c(0,6)) + 
    ggtitle(gsub('_',' ',sample), subtitle = subtitle) + theme_custom + 
    theme(aspect.ratio = length(unique(dt$id)) / length(sel), 
          strip.text = element_text(size = 12.8, face = 'plain', angle=0, vjust=0, hjust=0.5, margin = margin(5,0,5,0)), 
          axis.text.x = element_text(size = 12.8, color = 'black', angle = 90, vjust = 0.5, hjust = 1))
  if(!is.null(gr)) p + facet_grid(factor(condition, levels = levels(bd[[s]]$condition))~., scales = 'free', space = 'free') else p
}

# Plot expression
genes$sel <- c('Mki67','Ezh2','Id2','Id3','Gzma','Gzmb','Gzmk','Klrg1','Prf1','Fasl','Havcr2','Pdcd1','Lgals1','Lgals3','Il18r1','Il17ra','Cxcr3','Cxcr6','Klf2','Klf3','S1pr1','Tcf7','Slamf6','Ccr7','Bcl2','Il7r','Sell','Il6ra','Myc','Bcl2l1','Il2ra','Cd83','Tnfrsf18','Batf3','Bax','Icos','Cd81','Tnfrsf4','Tnfsf4','Tnfrsf9','Tnfsf14','Dhodh','Cad','Umps')
dp.sel <- dotplot2(object = bd[[s]], sel = genes$sel, gr=NULL) #ExtData10D

# Plot expression
df.bd <- data.frame(tx = bd[[s]]@reductions$tsne@cell.embeddings[,1], 
                    ty = bd[[s]]@reductions$tsne@cell.embeddings[,2], 
                    ux = bd[[s]]@reductions$umap@cell.embeddings[,1], 
                    uy = bd[[s]]@reductions$umap@cell.embeddings[,2], 
                    cells = colnames(bd[[s]]), 
                    sample = bd[[s]]@meta.data$sample, 
                    condition = bd[[s]]@meta.data$condition, 
                    cluster = bd[[s]]@meta.data$cluster, 
                    replicate = bd[[s]]@meta.data$replicate, 
                    t(as.matrix(GetAssayData(bd[[s]], slot = 'data')['Dhodh',,drop=F])))
df.bd <- reshape2::melt(df.bd, id.vars = c('tx', 'ty', 'ux', 'uy', 'cells', 'sample', 'condition','cluster','replicate'))
#df.bd$value.noise <- df.bd$value + rnorm(n = length(df.bd$value)) / 100000 #add noise

# Scale each gene
rescale <- function(scores,to=c(0,1)) if(any(scores > 1)) scales::rescale(scores, to=to) else scores
df.bd <- dplyr::bind_rows(Map(x=unique(df.bd$variable), function(x) {
  df.bd %>% filter(variable %in% x) %>% mutate(scaled = rescale(value))
}))

# Plot UMAP
um.Dhodh <- ggplot(df.bd %>% filter(variable %in% 'Dhodh'), aes(x = ux, y = uy, color = scaled)) + 
  ggrastr::rasterise(geom_point(shape = 16, size = 0.6, alpha = 1), dpi = 300) + 
  scale_color_gradientn('Exp.', colours = col$hm2, oob = scales::squish, 
                        limits = c(0,1), breaks = seq(0,1, length.out = 2), labels = c('min','max')) + 
  facet_grid(variable~condition) + 
  theme_custom2 + xlab('UMAP 1') + ylab('UMAP 2') + ggtitle(paste0('OT1 cells'), subtitle = 'Gene expression')

# Draw ridgeplot
rpa <- RidgePlot(bd[[s]], features = c('Dhodh'), ncol = 1, group.by = 'condition', assay = 'SCTmerged', slot = 'data', cols = c(alpha(col$gp[1], alpha = 0.6),col$gp[2]) )
```

### Transcriptional trajectory with diffusion pseudotime

- Use merged data

```{r trajectory, collapse=T, echo=F, warning=F, message=F, include=F}
# Extract normalized data from Seurat object
sd <- t(as.matrix(GetAssayData(bd[[s]], slot = "data")))

# Add annotation
sd <- destiny::as.ExpressionSet(as.data.frame(sd))
sd$cells <- names(Idents(bd[[s]]))
sd$condition <- bd[[s]]$condition
sd$sample <- bd[[s]]$sample
sd$cluster <- bd[[s]]$cluster

# Define variable genes to compute the diffusion map
varGenes <- as.character(VariableFeatures(bd[[s]]))

# Create diffusion map
rm(.Random.seed); .Random.seed <- as.integer(rep(1,626))
dm <- destiny::DiffusionMap(sd, verbose = T, n_eigs = 20, n_pcs = 50, vars = varGenes)
#saveRDS(dm, file = paste0(path, "/diffusion_map.rds"))
#dm <- readRDS(paste0(path, "/diffusion_map.rds"))

# Create pseudotime ordering
dpt <- destiny::DPT(dm, which.min(dm$DC1))
rm(.Random.seed)
#saveRDS(dpt, file = paste0(path, "/dpt.rds"))
#dpt <- readRDS(paste0(path, "/dpt.rds"))

# Extract data
df <- destiny::as.data.frame(dm)
df$DPT <- dpt[[paste0('DPT',which.min(dm$DC1))]]
df$tips <- ifelse(1:nrow(df) %in% destiny::tips(dpt)[1:3], 'tip', NA)
plot.dpt <- destiny::plot(dpt, pch = 20, axes = T, main = "dpt", dcs = c(1,3), root = 1, paths_to = c(2,3), w_width = 0.05, 
                          col_path = c("firebrick3", "royalblue2"), col_tip = "red", col_by = paste0('DPT',which.min(dm$DC1)) )

# Plot maps and pseudotime
plot.dm <- ggplot() + 
  geom_point(data=df, aes(DC1, DC3, colour = cluster), shape = 16, size = 0.8, alpha = 1) + 
  geom_path(data=as.data.frame(ggplot_build(plot.dpt)$data[3][[1]]), aes(x = x, y = y), size = 2, color = 'dodgerblue2') + 
  geom_path(data=as.data.frame(ggplot_build(plot.dpt)$data[4][[1]]), aes(x = x, y = y), size = 2, color = 'black') + 
  scale_color_manual("Cluster", values = col$cl, guide = guide_legend(override.aes = list(size = 4, alpha = 1))) + 
  scale_size_manual("", values = c(4,NA)) + ggtitle('OT1 cells', subtitle = 'Diffusion pseudotime') + 
  facet_grid(.~condition, scales = 'free', space = 'free') + theme_custom2 + theme(aspect.ratio = NULL)
```

-------------------------------------------------------------------------------------------------------------------

## Export plots

```{r filterdatasave, collapse=T, echo=F, warning=F, message=F, include=F}
# Save umap plots
ggsave(plot = grid(um.cls,rbind(c(1,1,1,1,1,1,1,1,1,2,3))), file = paste0(path, '/ExtData10A.pdf'), height = 4.32, width = 7.2)
ggsave(plot = grid(um.clr,rbind(c(1,1,1,1,1,1,1,1,1,2,3))), file = paste0(path, '/ExtData10B.pdf'), height = 6.6, width = 11.7)

# Save distribution plot
ggsave(plot = rmclip(dist.spclds), file = paste0(path, '/ExtData10C.pdf'), height = 3.8, width = 7)

# Save percentage expressing cells
ggsave(plot = rmclip(dp.sel), file = paste0(path, '/ExtData10D.pdf'), height = 4, width = 13)

# Save expression plots over UMAP
ggsave(plot = grid(um.Dhodh,rbind(c(1,1,1,1,1,2,3))), file = paste0(path, '/ExtData10E.pdf'), height = 4.5, width = 7.3)

# Save ridge plot
ggsave(plot = rmclip(rpa), file = paste0(path, '/ExtData10F.pdf'), height = 3, width = 5)

# Save Diffusion Map plots
ggsave(paste0(path, '/ExtData10G.pdf'), grid(plot.dm, rbind(c(1,1,1,1,1,1,1,1,2,2,3))), height = 4.35, width = 7.7)
```

-------------------------------------------------------------------------------------------------------------------

## Session info

```{r sessioninfo, collapse=T, echo=F, warning=F, message=F, include=F}
devtools::session_info()
#─ Session info ──────────────────────────────────────────────────────────────────────────────────────────────────────────────
# setting  value                       
# version  R version 4.0.4 (2021-02-15)
# os       macOS Big Sur 11.2.3        
# system   x86_64, darwin17.0          
# ui       RStudio                     
# language (EN)                        
# collate  en_US.UTF-8                 
# ctype    en_US.UTF-8                 
# tz       Europe/Berlin               

#─ Packages ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────
# package              * version   date       lib source                                    
# abind                  1.4-5     2016-07-21 [1] CRAN (R 4.0.2)                            
# assertthat             0.2.1     2019-03-21 [1] CRAN (R 4.0.2)                            
# beeswarm               0.3.1     2021-03-07 [1] CRAN (R 4.0.2)                            
# Biobase                2.50.0    2020-10-27 [1] Bioconductor                              
# BiocGenerics           0.36.0    2020-10-27 [1] Bioconductor                              
# BiocManager            1.30.10   2019-11-16 [1] CRAN (R 4.0.2)                            
# bitops                 1.0-6     2013-08-17 [1] CRAN (R 4.0.2)                            
# boot                   1.3-26    2021-01-25 [1] CRAN (R 4.0.4)                            
# cachem                 1.0.4     2021-02-13 [1] CRAN (R 4.0.2)                            
# Cairo                  1.5-12.2  2020-07-07 [1] CRAN (R 4.0.2)                            
# callr                  3.5.1     2020-10-13 [1] CRAN (R 4.0.2)                            
# car                    3.0-10    2020-09-29 [1] CRAN (R 4.0.2)                            
# carData                3.0-4     2020-05-22 [1] CRAN (R 4.0.2)                            
# cellranger             1.1.0     2016-07-27 [1] CRAN (R 4.0.2)                            
# circlize               0.4.12    2021-01-08 [1] CRAN (R 4.0.2)                            
# class                  7.3-18    2021-01-24 [1] CRAN (R 4.0.4)                            
# cli                    2.3.1     2021-02-23 [1] CRAN (R 4.0.2)                            
# cluster                2.1.0     2019-06-19 [1] CRAN (R 4.0.4)                            
# codetools              0.2-18    2020-11-04 [1] CRAN (R 4.0.4)                            
# colorspace             2.0-0     2020-11-11 [1] CRAN (R 4.0.2)                            
# cowplot                1.1.1     2020-12-30 [1] CRAN (R 4.0.2)                            
# crayon                 1.4.1     2021-02-08 [1] CRAN (R 4.0.2)                            
# curl                   4.3       2019-12-02 [1] CRAN (R 4.0.1)                            
# data.table             1.14.0    2021-02-21 [1] CRAN (R 4.0.2)                            
# DBI                    1.1.1     2021-01-15 [1] CRAN (R 4.0.2)                            
# DelayedArray           0.16.2    2021-02-26 [1] Bioconductor                              
# DelayedMatrixStats     1.12.3    2021-02-03 [1] Bioconductor                              
# deldir                 0.2-10    2021-02-16 [1] CRAN (R 4.0.2)                            
# DEoptimR               1.0-9     2021-05-24 [1] CRAN (R 4.0.2)                            
# desc                   1.3.0     2021-03-05 [1] CRAN (R 4.0.2)                            
# destiny                3.4.0     2020-10-27 [1] Bioconductor                              
# devtools               2.3.2     2020-09-18 [1] CRAN (R 4.0.2)                            
# digest                 0.6.27    2020-10-24 [1] CRAN (R 4.0.2)                            
# dplyr                * 1.0.5     2021-03-05 [1] CRAN (R 4.0.2)                            
# DT                     0.17      2021-01-06 [1] CRAN (R 4.0.2)                            
# e1071                  1.7-8     2021-07-28 [1] CRAN (R 4.0.2)                            
# ellipsis               0.3.1     2020-05-15 [1] CRAN (R 4.0.2)                            
# evaluate               0.14      2019-05-28 [1] CRAN (R 4.0.1)                            
# fansi                  0.4.2     2021-01-15 [1] CRAN (R 4.0.2)                            
# farver                 2.1.0     2021-02-28 [1] CRAN (R 4.0.2)                            
# fastmap                1.1.0     2021-01-25 [1] CRAN (R 4.0.2)                            
# fitdistrplus           1.1-3     2020-12-05 [1] CRAN (R 4.0.2)                            
# forcats                0.5.1     2021-01-27 [1] CRAN (R 4.0.2)                            
# foreign                0.8-81    2020-12-22 [1] CRAN (R 4.0.4)                            
# fs                     1.5.0     2020-07-31 [1] CRAN (R 4.0.2)                            
# future                 1.21.0    2020-12-10 [1] CRAN (R 4.0.2)                            
# future.apply           1.7.0     2021-01-04 [1] CRAN (R 4.0.2)                            
# generics               0.1.0     2020-10-31 [1] CRAN (R 4.0.2)                            
# GenomeInfoDb           1.26.4    2021-03-10 [1] Bioconductor                              
# GenomeInfoDbData       1.2.4     2021-03-15 [1] Bioconductor                              
# GenomicRanges          1.42.0    2020-10-27 [1] Bioconductor                              
# ggbeeswarm             0.6.0     2017-08-07 [1] CRAN (R 4.0.2)                            
# ggplot.multistats      1.0.0     2019-10-28 [1] CRAN (R 4.0.2)                            
# ggplot2              * 3.3.3     2020-12-30 [1] CRAN (R 4.0.2)                            
# ggrastr                0.2.3     2021-03-01 [1] CRAN (R 4.0.2)                            
# ggrepel                0.9.1     2021-01-15 [1] CRAN (R 4.0.2)                            
# ggridges               0.5.3     2021-01-08 [1] CRAN (R 4.0.2)                            
# ggthemes               4.2.4     2021-01-20 [1] CRAN (R 4.0.2)                            
# glmGamPoi              1.2.0     2020-10-27 [1] Bioconductor                              
# GlobalOptions          0.1.2     2020-06-10 [1] CRAN (R 4.0.2)                            
# globals                0.14.0    2020-11-22 [1] CRAN (R 4.0.2)                            
# glue                   1.4.2     2020-08-27 [1] CRAN (R 4.0.2)                            
# goftest                1.2-2     2019-12-02 [1] CRAN (R 4.0.2)                            
# gridExtra              2.3       2017-09-09 [1] CRAN (R 4.0.2)                            
# gtable                 0.3.0     2019-03-25 [1] CRAN (R 4.0.2)                            
# haven                  2.4.0     2021-04-14 [1] CRAN (R 4.0.2)                            
# hexbin                 1.28.2    2021-01-08 [1] CRAN (R 4.0.2)                            
# hms                    1.0.0     2021-01-13 [1] CRAN (R 4.0.2)                            
# htmltools              0.5.4     2022-12-07 [1] CRAN (R 4.0.4)                            
# htmlwidgets            1.5.3     2020-12-10 [1] CRAN (R 4.0.2)                            
# httpuv                 1.5.5     2021-01-13 [1] CRAN (R 4.0.2)                            
# httr                   1.4.2     2020-07-20 [1] CRAN (R 4.0.2)                            
# ica                    1.0-2     2018-05-24 [1] CRAN (R 4.0.2)                            
# igraph                 1.2.6     2020-10-06 [1] CRAN (R 4.0.2)                            
# IRanges                2.24.1    2020-12-12 [1] Bioconductor                              
# irlba                  2.3.3     2019-02-05 [1] CRAN (R 4.0.2)                            
# jsonlite               1.7.2     2020-12-09 [1] CRAN (R 4.0.2)                            
# KernSmooth             2.23-18   2020-10-29 [1] CRAN (R 4.0.4)                            
# knitr                  1.36      2021-09-29 [1] CRAN (R 4.0.2)                            
# knn.covertree          1.0       2019-10-28 [1] CRAN (R 4.0.2)                            
# labeling               0.4.2     2020-10-20 [1] CRAN (R 4.0.2)                            
# laeken                 0.5.2     2021-10-06 [1] CRAN (R 4.0.2)                            
# later                  1.1.0.1   2020-06-05 [1] CRAN (R 4.0.2)                            
# lattice                0.20-41   2020-04-02 [1] CRAN (R 4.0.4)                            
# lazyeval               0.2.2     2019-03-15 [1] CRAN (R 4.0.2)                            
# leiden                 0.3.7     2021-01-26 [1] CRAN (R 4.0.2)                            
# lifecycle              1.0.0     2021-02-15 [1] CRAN (R 4.0.2)                            
# listenv                0.8.0     2019-12-05 [1] CRAN (R 4.0.2)                            
# lmtest                 0.9-38    2020-09-09 [1] CRAN (R 4.0.2)                            
# magrittr               2.0.1     2020-11-17 [1] CRAN (R 4.0.2)                            
# MASS                   7.3-53    2020-09-09 [1] CRAN (R 4.0.4)                            
# Matrix                 1.3-2     2021-01-06 [1] CRAN (R 4.0.4)                            
# MatrixGenerics         1.2.1     2021-01-30 [1] Bioconductor                              
# matrixStats            0.58.0    2021-01-29 [1] CRAN (R 4.0.2)                            
# memoise                2.0.1     2021-11-26 [1] CRAN (R 4.0.2)                            
# mgcv                   1.8-33    2020-08-27 [1] CRAN (R 4.0.4)                            
# mime                   0.10      2021-02-13 [1] CRAN (R 4.0.2)                            
# miniUI                 0.1.1.1   2018-05-18 [1] CRAN (R 4.0.2)                            
# munsell                0.5.0     2018-06-12 [1] CRAN (R 4.0.2)                            
# nlme                   3.1-152   2021-02-04 [1] CRAN (R 4.0.4)                            
# nnet                   7.3-15    2021-01-24 [1] CRAN (R 4.0.4)                            
# openxlsx               4.2.3     2020-10-27 [1] CRAN (R 4.0.2)                            
# parallelly             1.24.0    2021-03-14 [1] CRAN (R 4.0.4)                            
# patchwork              1.1.1     2020-12-17 [1] CRAN (R 4.0.2)                            
# pbapply                1.4-3     2020-08-18 [1] CRAN (R 4.0.2)                            
# pcaMethods             1.82.0    2020-10-27 [1] Bioconductor                              
# pheatmap               1.0.12    2019-01-04 [1] CRAN (R 4.0.2)                            
# pillar                 1.5.1     2021-03-05 [1] CRAN (R 4.0.2)                            
# pkgbuild               1.2.0     2020-12-15 [1] CRAN (R 4.0.2)                            
# pkgconfig              2.0.3     2019-09-22 [1] CRAN (R 4.0.2)                            
# pkgload                1.2.0     2021-02-23 [1] CRAN (R 4.0.2)                            
# plotly                 4.9.3     2021-01-10 [1] CRAN (R 4.0.2)                            
# plyr                   1.8.6     2020-03-03 [1] CRAN (R 4.0.2)                            
# png                    0.1-7     2013-12-03 [1] CRAN (R 4.0.2)                            
# polyclip               1.10-0    2019-03-14 [1] CRAN (R 4.0.2)                            
# prettyunits            1.1.1     2020-01-24 [1] CRAN (R 4.0.2)                            
# processx               3.5.0     2021-03-23 [1] CRAN (R 4.0.4)                            
# promises               1.2.0.1   2021-02-11 [1] CRAN (R 4.0.2)                            
# proxy                  0.4-26    2021-06-07 [1] CRAN (R 4.0.2)                            
# ps                     1.6.0     2021-02-28 [1] CRAN (R 4.0.2)                            
# purrr                  0.3.4     2020-04-17 [1] CRAN (R 4.0.2)                            
# R.methodsS3            1.8.1     2020-08-26 [1] CRAN (R 4.0.2)                            
# R.oo                   1.24.0    2020-08-26 [1] CRAN (R 4.0.2)                            
# R.utils                2.10.1    2020-08-26 [1] CRAN (R 4.0.2)                            
# R6                     2.5.0     2020-10-28 [1] CRAN (R 4.0.2)                            
# ranger                 0.13.1    2021-07-14 [1] CRAN (R 4.0.2)                            
# RANN                   2.6.1     2019-01-08 [1] CRAN (R 4.0.2)                            
# RColorBrewer           1.1-2     2014-12-07 [1] CRAN (R 4.0.2)                            
# Rcpp                   1.0.8.3   2022-03-17 [1] CRAN (R 4.0.5)                            
# RcppAnnoy              0.0.18    2020-12-15 [1] CRAN (R 4.0.2)                            
# RcppEigen              0.3.3.9.1 2020-12-17 [1] CRAN (R 4.0.2)                            
# RcppHNSW               0.3.0     2020-09-06 [1] CRAN (R 4.0.2)                            
# RCurl                  1.98-1.3  2021-03-16 [1] CRAN (R 4.0.4)                            
# readxl                 1.3.1     2019-03-13 [1] CRAN (R 4.0.2)                            
# remotes                2.2.0     2020-07-21 [1] CRAN (R 4.0.2)                            
# reshape2               1.4.4     2020-04-09 [1] CRAN (R 4.0.2)                            
# reticulate             1.25      2022-05-11 [1] CRAN (R 4.0.4)                            
# rio                    0.5.26    2021-03-01 [1] CRAN (R 4.0.2)                            
# rlang                  0.4.10    2020-12-30 [1] CRAN (R 4.0.2)                            
# rmarkdown              2.19      2022-12-15 [1] CRAN (R 4.0.4)                            
# robustbase             0.93-9    2021-09-27 [1] CRAN (R 4.0.2)                            
# ROCR                   1.0-11    2020-05-02 [1] CRAN (R 4.0.2)                            
# rpart                  4.1-15    2019-04-12 [1] CRAN (R 4.0.4)                            
# rprojroot              2.0.2     2020-11-15 [1] CRAN (R 4.0.2)                            
# RSpectra               0.16-0    2019-12-01 [1] CRAN (R 4.0.2)                            
# rstudioapi             0.13      2020-11-12 [1] CRAN (R 4.0.2)                            
# rsvd                   1.0.5     2021-04-16 [1] CRAN (R 4.0.2)                            
# Rtsne                  0.15      2018-11-10 [1] CRAN (R 4.0.2)                            
# S4Vectors              0.28.1    2020-12-09 [1] Bioconductor                              
# scales                 1.1.1     2020-05-11 [1] CRAN (R 4.0.2)                            
# scattermore            0.7       2020-11-24 [1] CRAN (R 4.0.2)                            
# scatterplot3d          0.3-41    2018-03-14 [1] CRAN (R 4.0.2)                            
# sctransform          * 0.3.2     2020-12-16 [1] CRAN (R 4.0.2)                            
# sessioninfo            1.1.1     2018-11-05 [1] CRAN (R 4.0.2)                            
# Seurat               * 4.0.1     2021-03-18 [1] CRAN (R 4.0.4)                            
# SeuratObject         * 4.0.0     2021-01-15 [1] CRAN (R 4.0.2)                            
# SeuratWrappers         0.3.0     2021-08-03 [1] Github (satijalab/seurat-wrappers@8510069)
# shape                  1.4.5     2020-09-13 [1] CRAN (R 4.0.2)                            
# shiny                  1.6.0     2021-01-25 [1] CRAN (R 4.0.2)                            
# SingleCellExperiment   1.12.0    2020-10-27 [1] Bioconductor                              
# smoother               1.1       2015-04-16 [1] CRAN (R 4.0.2)                            
# sp                     1.4-5     2021-01-10 [1] CRAN (R 4.0.2)                            
# sparseMatrixStats      1.2.1     2021-02-02 [1] Bioconductor                              
# spatstat.core          2.0-0     2021-03-23 [1] CRAN (R 4.0.4)                            
# spatstat.data          2.1-0     2021-03-21 [1] CRAN (R 4.0.2)                            
# spatstat.geom          2.0-1     2021-03-22 [1] CRAN (R 4.0.2)                            
# spatstat.sparse        2.0-0     2021-03-16 [1] CRAN (R 4.0.4)                            
# spatstat.utils         2.1-0     2021-03-15 [1] CRAN (R 4.0.4)                            
# stringi                1.5.3     2020-09-09 [1] CRAN (R 4.0.2)                            
# stringr                1.4.0     2019-02-10 [1] CRAN (R 4.0.2)                            
# SummarizedExperiment   1.20.0    2020-10-27 [1] Bioconductor                              
# survival               3.2-7     2020-09-28 [1] CRAN (R 4.0.4)                            
# tensor                 1.5       2012-05-05 [1] CRAN (R 4.0.2)                            
# testthat               3.0.2     2021-02-14 [1] CRAN (R 4.0.2)                            
# tibble                 3.1.0     2021-02-25 [1] CRAN (R 4.0.2)                            
# tidyr                  1.1.3     2021-03-03 [1] CRAN (R 4.0.2)                            
# tidyselect             1.1.0     2020-05-11 [1] CRAN (R 4.0.2)                            
# trqwe                  0.1       2021-08-11 [1] Github (traversc/trqwe@dcf5767)           
# TTR                    0.24.2    2020-09-01 [1] CRAN (R 4.0.2)                            
# usethis                2.0.1     2021-02-10 [1] CRAN (R 4.0.2)                            
# utf8                   1.2.1     2021-03-12 [1] CRAN (R 4.0.2)                            
# uwot                   0.1.10    2020-12-15 [1] CRAN (R 4.0.2)                            
# vcd                    1.4-8     2020-09-21 [1] CRAN (R 4.0.2)                            
# vctrs                  0.3.6     2020-12-17 [1] CRAN (R 4.0.2)                            
# VIM                    6.1.1     2021-07-22 [1] CRAN (R 4.0.2)                            
# vipor                  0.4.5     2017-03-22 [1] CRAN (R 4.0.2)                            
# viridisLite            0.3.0     2018-02-01 [1] CRAN (R 4.0.1)                            
# withr                  2.4.1     2021-01-26 [1] CRAN (R 4.0.2)                            
# xfun                   0.36      2022-12-21 [1] CRAN (R 4.0.4)                            
# xtable                 1.8-4     2019-04-21 [1] CRAN (R 4.0.2)                            
# xts                    0.12.1    2020-09-09 [1] CRAN (R 4.0.2)                            
# XVector                0.30.0    2020-10-28 [1] Bioconductor                              
# yaml                   2.2.1     2020-02-01 [1] CRAN (R 4.0.2)                            
# zip                    2.1.1     2020-08-27 [1] CRAN (R 4.0.2)                            
# zlibbioc               1.36.0    2020-10-28 [1] Bioconductor                              
# zoo                    1.8-9     2021-03-09 [1] CRAN (R 4.0.2)   
```

-------------------------------------------------------------------------------------------------------------------
