---
title: "bulk"
output: html_document
---

## Prepare session

```{r setup, include=FALSE}
require(knitr)
knitr::opts_chunk$set(echo = TRUE)
path <- '~/Desktop/bulk'
dir.create(path = path, recursive = T)
knitr::opts_knit$set(root.dir = path, collapse = T, comment = '#>')
```

```{r loadlibraries, echo=FALSE}
library(DESeq2)
library(tibble)
library(gplots)
library(openxlsx)
library(dplyr)
library(limma)

# Define custom ggplot theme
theme_custom <- theme(axis.text.x = element_text(size = 12.8, color = 'black'),
                      axis.text.y = element_text(size = 12.8, color = 'black'), 
                      axis.title.x = element_text(size = 16, margin = margin(6,0,0,0)), 
                      axis.title.y = element_text(size = 16, margin = margin(0,8,0,0)),
                      plot.title = element_text(size = 16, face = 'plain', hjust = 0.5, margin = margin(5,0,10,0)),
                      plot.subtitle = element_text(size = 12, hjust = 0.5, margin = margin(0,0,10,0)), 
                      axis.ticks.x = element_line(size = 0.4, colour = 'black'), 
                      axis.ticks.y = element_line(size = 0.4, colour = 'black'),
                      axis.line = element_blank(), 
                      panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank(),
                      panel.grid.minor.x = element_blank(), panel.grid.major.x = element_blank(),
                      panel.border = element_rect(colour = 'black', size = 0.8, fill = NA),
                      panel.background = element_rect(colour = NA, fill = NA), 
                      strip.text.x = element_text(size = 16, face = 'plain', margin = margin(0,0,5,0) ), 
                      strip.text.y = element_text(size = 16, face = 'plain', margin = margin(0,0,0,5) ), 
                      strip.background = element_blank(), 
                      legend.key = element_blank(), legend.key.size = unit(0.5,'cm'),
                      legend.text = element_text(size = 12.8), 
                      legend.title = element_text(size = 12.8),
                      plot.background = element_rect(colour = NA, fill = NA), aspect.ratio = 1)
```

-------------------------------------------------------------------------------------------------------------------

## DEG function

```{r DEGfunction, echo=FALSE}
Deseq.Func <- function(ReadCountTable, DesignTable, Filename, filepath, Treat, Ctr, 
                       BaseMean= 50, logFC = 1, Padj = 0.05){
  
  #gene pre-filtering
  ReadCountTable <- ReadCountTable[rowSums(ReadCountTable) > 10,]
  
  #perform DEGA
  dds <- DESeqDataSetFromMatrix(countData = ReadCountTable,
                                colData = DesignTable,
                                design = ~ Run + Treatment)
  dds <- DESeq(dds)
  resultsNames(dds)
  
  #get raw DEG list
  res <- as.data.frame(results(dds, contrast = c("Treatment", Treat, Ctr)))
  res$geneName <- rownames(res)
  
  # export norm/rlog, rawDEG data
  write.table(x = rownames_to_column(as.data.frame(counts(dds, normalized = T)),'Gene'),
            file = paste0(filepath,'Norm_', Filename, '.csv'), sep = '\t', row.names = F, quote = F)

  # time consuming
  write.table(x = rownames_to_column(as.data.frame(assay(rlog(dds, blind = F))),'Gene'),
            file = paste0(filepath,'Rlog_', Filename,'.csv'), sep = '\t', row.names = F, quote = F)
  write.xlsx(x = na.omit(res), file = paste0(filepath, Filename, '_', 'Raw_DEG_', Treat, '_VS_', Ctr, '.xlsx'))
  write.xlsx(x = na.omit(res[which(res$baseMean >= BaseMean & abs(res$log2FoldChange) >= logFC &
                                     res$padj < Padj),]), file = paste0(filepath, Filename, '_', 'Filter_DEG_',
                                                                        Treat, '_VS_', Ctr,'.xlsx'))
}
```

-------------------------------------------------------------------------------------------------------------------

## DEG test

```{r}
# Download pre-processed data from GEO SuperSeries and place at '~/Desktop/bulk/'
download.file(url='https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE200358&format=file&file=GSE200358%5FCombined%5FRun2%2B3%2B17%2B19Raw%2Etxt%2Egz', destfile=paste0(path, '/GSE200358_Combined_Run2+3+17+19Raw.txt.gz'))
download.file(url='https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE200358&format=file&file=GSE200358%5FDesign%5FCombined%5FRun2%2B3%2B17%2B19%2Etxt%2Egz', destfile=paste0(path, '/GSE200358_Design_Combined_Run2+3+17+19.txt.gz'))

# Prepare data
Expr = read.csv(gzfile(paste0(path, '/GSE200358_Combined_Run2+3+17+19Raw.txt.gz')), sep = '\t', header = T, row.names = 1, stringsAsFactors = F, check.names = F)
Design = read.csv(gzfile(paste0(path, '/GSE200358_Design_Combined_Run2+3+17+19.txt.gz')), sep = '\t', header = T, nrows = ncol(Expr), row.names = 2, stringsAsFactors = F, blank.lines.skip = T)[,-1] %>% rownames_to_column(., 'Sample')
#make sure identical(colnames(ReadCountTable), Design$sample) = True
Run2_Run3_D7_Design <- Design %>% filter(Day == 'Day7' & Run %in% c('Run2', 'Run3'))
Run2_Run3_D7_Expr <- Expr[, Run2_Run3_D7_Design$Sample]
Run17_Run19_D7_Design <- Design %>% filter(Day == 'Day7' & Run %in% c('Run17', 'Run19'))
Run17_Run19_D7_Expr <- Expr[, Run17_Run19_D7_Design$Sample]

# Find DEGs
Deseq.Func(ReadCountTable = Run2_Run3_D7_Expr, DesignTable = Run2_Run3_D7_Design, Filename = 'Run2_Run3_D7', filepath = paste0(path,'/'), Treat = 'TFL', Ctr = 'CMC')
Deseq.Func(ReadCountTable = Run17_Run19_D7_Expr, DesignTable = Run17_Run19_D7_Design, Filename = 'Run17_Run19_D7', filepath = paste0(path,'/'), Treat = 'TFL', Ctr = 'CMC')
```

-------------------------------------------------------------------------------------------------------------------

## Volcano plot

```{r}
library(ggplot2)

# Function to draw volcano plot
volcano <-  function(data, gene_highlight, title) {
  data$DE <- 'ns'
  data$DE <- ifelse(data$padj < 0.05 & data$log2FoldChange > 1.5, 'up', data$DE)
  data$DE <- ifelse(data$padj < 0.05 & data$log2FoldChange < -1.5, 'dn', data$DE)
  data$DE <- factor(data$DE, levels = sort(unique(data$DE))[c(3,1,2)])
  data$DEf <- 'no'
  data$DEf <- ifelse(data$gene %in% gene_highlight & data$log2FoldChange > 0, 'up', data$DEf)
  data$DEf <- ifelse(data$gene %in% gene_highlight & data$log2FoldChange < 0, 'dn', data$DEf)
  data$DEf <- factor(data$DEf, levels = c('up','dn','ns','no'))
  data$DEs <- 'no'
  #data$DEs <- ifelse(-log10(data$padj) > 10 | data$LFC > 10, 'up', data$DEs)
  #data$DEs <- ifelse(-log10(data$padj) > 10 | data$LFC < -10, 'dn', data$DEs)
  data$DEs <- ifelse(data$padj < 1e-30 | data$log2FoldChange > 5, 'up', data$DEs)
  data$DEs <- ifelse(data$padj < 1e-30 | data$log2FoldChange < -5, 'dn', data$DEs)
  data$DEs <- factor(data$DEs, levels = c('up','dn','no'))
  
  p <- ggplot(data, aes(x = ifelse(log2FoldChange>5,5,ifelse(log2FoldChange<(-5),-5,log2FoldChange)), y = ifelse(padj<1e-30,1e-30, padj) )) + #ifelse(-log10(padj)>10,10, -log10(padj))) ) + 
    #geom_point(aes(), alpha = 1, size = 1, shape = 21) +
    ggrastr::geom_point_rast(aes(color = DE, fill = DEf, size = DEf, shape = DEs), 
                             alpha = 1, stroke = 0.2, raster.dpi = 600) + 
    annotate('text', x = 4, y = 0.5, label = paste0(sum(data$padj < 0.05 & data$log2FoldChange > 1.5), ' up'), size = 5) + 
    annotate('text', x = -4, y = 0.5, label = paste0(sum(data$padj < 0.05 & data$log2FoldChange < -1.5), ' dn'), size = 5) + 
    geom_vline(xintercept = c(-1.5,1.5), colour = "black", linetype = "dashed", size = 0.3) + 
    geom_hline(yintercept = c(0.05), colour = "firebrick3", linetype = "dashed", size = 0.3) + 
    scale_x_continuous(expression(""*log[2]*"(fold change)"), limits = c(-5,5), breaks = seq(-5,5,1), expand = c(0.02,0)) + 
    #scale_y_continuous(expression("-"*log[10]*"(p-value)"), limits = c(0,10), breaks = seq(0,10,2), expand = c(0.02,0)) + 
    #scale_y_log10(expression("Adj. p-value (in "*log[10]*" scale)"), labels = scales::trans_format("log10", scales::math_format(10^.x)) ) + 
    scale_y_continuous(expression("Adj. p-value (in "*log[10]*" scale)"), trans = ggforce::trans_reverser('log10'), limits = c(1,1e-30), breaks = c(1,1e-10,1e-20,1e-30), labels = scales::trans_format("log10", scales::math_format(10^.x)), expand = c(0.02,0)) + #annotation_logticks(base=10, sides = 'l') + 
    #scale_fill_manual("Regulation", values = c('grey60', 'white'), #c("black", "grey60", "white"), 
    #                  guide = guide_legend(override.aes = list(size = 5, alpha = 1))) + 
    scale_size_manual("Regulation", values = c(up=2,dn=2,ns=2,no=1), guide = F) + 
    scale_shape_manual("Regulation", values = c(up=24,dn=24,ns=24,no=21), guide = F) + 
    scale_fill_manual("Regulation", values = c(up='#B2182B', dn='#2166AC', ns='grey70', no=NA), guide = F) + 
    scale_color_manual("Regulation", values = c(up='#B2182B',dn='#2166AC',ns='grey70'), 
                       guide = guide_legend(override.aes = list(size = 5, shape = 16, alpha = 1))) + 
    ggtitle(title) + 
    theme_custom
  p <- p + 
    ggrepel::geom_text_repel(data = subset(data, geneName %in% gene_highlight & log2FoldChange > 0), segment.size=0.2, seed = 1, aes(label = geneName), color = "grey30", size = 3, min.segment.length = 0, force = 5, direction = 'y', nudge_x = 5 - subset(data, geneName %in% gene_highlight & log2FoldChange > 0)$log2FoldChange, hjust=0, max.overlaps=20, box.padding = unit(0.2, "lines"), point.padding = unit(0.4, "lines"), fontface = "italic") + 
    ggrepel::geom_text_repel(data = subset(data, geneName %in% gene_highlight & log2FoldChange < 0), segment.size=0.2, seed = 1, aes(label = geneName), color = "grey30", size = 3, min.segment.length = 0, force = 5, direction = 'y', nudge_x = -5 - subset(data, geneName %in% gene_highlight & log2FoldChange < 0)$log2FoldChange, hjust=1, max.overlaps=20, box.padding = unit(0.2, "lines"), point.padding = unit(0.4, "lines"), fontface = "italic")
  return(p)
}

# Create volcano plots
vp1 <- volcano(data=read.xlsx(paste0(path, '/Run2_Run3_D7_Raw_DEG_TFL_VS_CMC.xlsx')) %>% filter(baseMean > 50), 
               gene_highlight = c('Esm1','Klra4','Klrc1','Klre1','Plekha6','Itga1','Gm4208','S1pr5','Klra9','Klrb1c',' Pdcd1','Cenpk','Shmt1','Ccr7','Capn3','Tox','Izumo1r','Bcat1','Ly6i','Lpl','Saa3','Gbp2b'), title = 'Total OT-1 (day 7)'
)
vp2 <- volcano(data=read.xlsx(paste0(path, '/Run17_Run19_D7_Raw_DEG_TFL_VS_CMC.xlsx')) %>% filter(baseMean > 50), 
               gene_highlight = c('Gm4208','Pls1','Klrb1c','Cx3cr1','Mal','Cd163l1','Plekha6','Epha3',' Tnfsf8','Sell','St8sia1','Capg','Capn3','Cd200','Eng','Phlda3','Serpine2','Tox2','Bcat1'), title = 'CD127+KLRG1- OT-1 (day 7)'
)

# Save volcano plots
ggsave(paste0(path, '/Fig3A.pdf'), vp1, height = 4.5, width = 6)
ggsave(paste0(path, '/Fig3C.pdf'), vp2, height = 4.5, width = 6)
```

-------------------------------------------------------------------------------------------------------------------

## Gene expression

```{r}
library(pheatmap)
library(RColorBrewer)
library(openxlsx)

# Prepare data
Norm_df = read.csv(paste0(path, '/Rlog_Run2_Run3_D7.csv'), sep = '\t', header = T, row.names = 1)
Norm_zscore = t(scale(t(Norm_df[, c(6:10, 1:5)])))
new_annotation = data.frame(row.names = colnames(Norm_zscore),
                            condition = as.factor(c(rep('CMC', 5), rep('TFL',5))), stringsAsFactors = F)
new_annotation_color = list(condition = c('CMC' = '#BF4342', 'TFL' = '#033F63'))

# Plot expression over heatmap
heat <- pheatmap(Norm_zscore[match(toupper(Gene_highlight), toupper(rownames(Norm_zscore))),], cluster_rows = F, cluster_cols = F, show_colnames = F, annotation = new_annotation, cutree_cols = 1, gaps_col = c(5), annotation_colors = new_annotation_color, cellheight = 10, cellwidth = 10, fontsize = 10, border_color = 'white', legend_breaks = seq(-1.5,1.5,0.5), breaks = seq(-1.5,1.5,length.out=42), main = 'Total OT-1 (day 7)', color = colorRampPalette(c('#133d67','#2166AC','#90b2d5','#d2e0ee','grey90','#efd0d4','#d88b95','#B2182B','#590c15'))(41))

# Save heatmap
ggsave(paste0(path, '/Fig3B.pdf'), heat[[4]], height = 4, width = 4)
```

-------------------------------------------------------------------------------------------------------------------

## PCA plot

```{r}
library(ggfortify)

# Prepare data
Run2_Run3_D7_D28_Design = Design %>% filter(Run %in% c('Run2', 'Run3'))
Run2_Run3_D7_D28_Expr = Expr[, Run2_Run3_D7_D28_Design$Sample]
Run2_Run3_D7_D28_Design$group = paste0(Run2_Run3_D7_D28_Design$Day, '_', Run2_Run3_D7_D28_Design$Treatment)
Run2_Run3_D7_D28_Design$cluster = plyr::mapvalues(Run2_Run3_D7_D28_Design$group, from = unique(Run2_Run3_D7_D28_Design$group), to = c(1,2,3,4))

# run deseq2
dds <- DESeqDataSetFromMatrix(countData = Run2_Run3_D7_D28_Expr, colData = Run2_Run3_D7_D28_Design, design = ~ cluster)
rlogcounts = as.data.frame(assay(rlog(dds, blind = FALSE)))

# remove batch effects
rlogcounts <- limma::removeBatchEffect(rlogcounts, as.factor(dds$Run))
rlogcounts <- rlogcounts[apply(rlogcounts, 1, sd) > quantile(apply(rlogcounts, 1, sd), probs=seq(0,1,0.001)[990]), ]

# run PCA
pcDat <- prcomp(t(rlogcounts))

# plot PCA
pca <- autoplot(cluster::fanny(pcDat$x, 4), frame = TRUE, frame.type = 'norm', data = Run2_Run3_D7_D28_Design, label_size = 4) + 
  geom_point(aes(color = cluster), size = 4) + theme_custom + scale_y_reverse() + scale_x_reverse() + 
  scale_color_manual(values=c('1'='#50b2c1', '2'='grey50', '3'='#1266b1', '4'='black'), labels=c(sub('_',' ',unique(Run2_Run3_D7_D28_Design$group))), guide = guide_legend(reverse=T) ) + 
  scale_fill_manual(values=c('1'='#50b2c1', '2'='grey50', '3'='#1485ea', '4'='black'), labels=c(sub('_',' ',unique(Run2_Run3_D7_D28_Design$group))), guide = guide_legend(reverse=T) )

# Save PCA plot
ggsave(paste0(path, '/Fig3D.pdf'), pca, height = 4, width = 6)
```

-------------------------------------------------------------------------------------------------------------------
## Session info

```{r sessioninfo, collapse=T, echo=F, warning=F, message=F, include=F}
devtools::session_info()
#─ Session info ─────────────────────────────────────────────────────────────────────────────────────────────────────────────
# setting  value                       
# version  R version 4.0.4 (2021-02-15)
# os       macOS Big Sur 11.2.3        
# system   x86_64, darwin17.0          
# ui       RStudio                     
# language (EN)                        
# collate  en_US.UTF-8                 
# ctype    en_US.UTF-8                 
# tz       Europe/Berlin               

#─ Packages ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────
# package              * version  date       lib source        
# annotate               1.68.0   2020-10-27 [1] Bioconductor  
# AnnotationDbi          1.52.0   2020-10-27 [1] Bioconductor  
# assertthat             0.2.1    2019-03-21 [1] CRAN (R 4.0.2)
# beeswarm               0.3.1    2021-03-07 [1] CRAN (R 4.0.2)
# Biobase              * 2.50.0   2020-10-27 [1] Bioconductor  
# BiocGenerics         * 0.36.0   2020-10-27 [1] Bioconductor  
# BiocParallel           1.24.1   2020-11-06 [1] Bioconductor  
# bit                    4.0.4    2020-08-04 [1] CRAN (R 4.0.2)
# bit64                  4.0.5    2020-08-30 [1] CRAN (R 4.0.2)
# bitops                 1.0-6    2013-08-17 [1] CRAN (R 4.0.2)
# blob                   1.2.1    2020-01-20 [1] CRAN (R 4.0.2)
# cachem                 1.0.4    2021-02-13 [1] CRAN (R 4.0.2)
# callr                  3.5.1    2020-10-13 [1] CRAN (R 4.0.2)
# caTools                1.18.1   2021-01-11 [1] CRAN (R 4.0.2)
# cli                    2.3.1    2021-02-23 [1] CRAN (R 4.0.2)
# cluster                2.1.0    2019-06-19 [1] CRAN (R 4.0.4)
# colorspace             2.0-0    2020-11-11 [1] CRAN (R 4.0.2)
# crayon                 1.4.1    2021-02-08 [1] CRAN (R 4.0.2)
# DBI                    1.1.1    2021-01-15 [1] CRAN (R 4.0.2)
# DelayedArray           0.16.2   2021-02-26 [1] Bioconductor  
# desc                   1.3.0    2021-03-05 [1] CRAN (R 4.0.2)
# DESeq2               * 1.30.1   2021-02-19 [1] Bioconductor  
# devtools               2.3.2    2020-09-18 [1] CRAN (R 4.0.2)
# digest                 0.6.27   2020-10-24 [1] CRAN (R 4.0.2)
# dplyr                * 1.0.5    2021-03-05 [1] CRAN (R 4.0.2)
# ellipsis               0.3.1    2020-05-15 [1] CRAN (R 4.0.2)
# evaluate               0.14     2019-05-28 [1] CRAN (R 4.0.1)
# fansi                  0.4.2    2021-01-15 [1] CRAN (R 4.0.2)
# farver                 2.1.0    2021-02-28 [1] CRAN (R 4.0.2)
# fastmap                1.1.0    2021-01-25 [1] CRAN (R 4.0.2)
# fs                     1.5.0    2020-07-31 [1] CRAN (R 4.0.2)
# genefilter             1.72.1   2021-01-21 [1] Bioconductor  
# geneplotter            1.68.0   2020-10-27 [1] Bioconductor  
# generics               0.1.0    2020-10-31 [1] CRAN (R 4.0.2)
# GenomeInfoDb         * 1.26.4   2021-03-10 [1] Bioconductor  
# GenomeInfoDbData       1.2.4    2021-03-15 [1] Bioconductor  
# GenomicRanges        * 1.42.0   2020-10-27 [1] Bioconductor  
# ggbeeswarm             0.6.0    2017-08-07 [1] CRAN (R 4.0.2)
# ggforce                0.3.3    2021-03-05 [1] CRAN (R 4.0.2)
# ggfortify            * 0.4.15   2022-11-11 [1] CRAN (R 4.0.4)
# ggplot2              * 3.3.3    2020-12-30 [1] CRAN (R 4.0.2)
# ggrastr                0.2.3    2021-03-01 [1] CRAN (R 4.0.2)
# ggrepel                0.9.1    2021-01-15 [1] CRAN (R 4.0.2)
# glue                   1.4.2    2020-08-27 [1] CRAN (R 4.0.2)
# gplots               * 3.1.1    2020-11-28 [1] CRAN (R 4.0.2)
# gridExtra              2.3      2017-09-09 [1] CRAN (R 4.0.2)
# gtable                 0.3.0    2019-03-25 [1] CRAN (R 4.0.2)
# gtools                 3.8.2    2020-03-31 [1] CRAN (R 4.0.2)
# htmltools              0.5.4    2022-12-07 [1] CRAN (R 4.0.4)
# httr                   1.4.2    2020-07-20 [1] CRAN (R 4.0.2)
# IRanges              * 2.24.1   2020-12-12 [1] Bioconductor  
# KernSmooth             2.23-18  2020-10-29 [1] CRAN (R 4.0.4)
# knitr                  1.36     2021-09-29 [1] CRAN (R 4.0.2)
# lattice                0.20-41  2020-04-02 [1] CRAN (R 4.0.4)
# lifecycle              1.0.0    2021-02-15 [1] CRAN (R 4.0.2)
# limma                * 3.46.0   2020-10-27 [1] Bioconductor  
# locfit                 1.5-9.4  2020-03-25 [1] CRAN (R 4.0.2)
# magrittr               2.0.1    2020-11-17 [1] CRAN (R 4.0.2)
# MASS                   7.3-53   2020-09-09 [1] CRAN (R 4.0.4)
# Matrix                 1.3-2    2021-01-06 [1] CRAN (R 4.0.4)
# MatrixGenerics       * 1.2.1    2021-01-30 [1] Bioconductor  
# matrixStats          * 0.58.0   2021-01-29 [1] CRAN (R 4.0.2)
# memoise                2.0.1    2021-11-26 [1] CRAN (R 4.0.2)
# munsell                0.5.0    2018-06-12 [1] CRAN (R 4.0.2)
# openxlsx             * 4.2.3    2020-10-27 [1] CRAN (R 4.0.2)
# pheatmap             * 1.0.12   2019-01-04 [1] CRAN (R 4.0.2)
# pillar                 1.5.1    2021-03-05 [1] CRAN (R 4.0.2)
# pkgbuild               1.2.0    2020-12-15 [1] CRAN (R 4.0.2)
# pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.0.2)
# pkgload                1.2.0    2021-02-23 [1] CRAN (R 4.0.2)
# plyr                   1.8.6    2020-03-03 [1] CRAN (R 4.0.2)
# polyclip               1.10-0   2019-03-14 [1] CRAN (R 4.0.2)
# prettyunits            1.1.1    2020-01-24 [1] CRAN (R 4.0.2)
# processx               3.5.0    2021-03-23 [1] CRAN (R 4.0.4)
# ps                     1.6.0    2021-02-28 [1] CRAN (R 4.0.2)
# purrr                  0.3.4    2020-04-17 [1] CRAN (R 4.0.2)
# R6                     2.5.0    2020-10-28 [1] CRAN (R 4.0.2)
# RColorBrewer         * 1.1-2    2014-12-07 [1] CRAN (R 4.0.2)
# Rcpp                   1.0.8.3  2022-03-17 [1] CRAN (R 4.0.5)
# RCurl                  1.98-1.3 2021-03-16 [1] CRAN (R 4.0.4)
# remotes                2.2.0    2020-07-21 [1] CRAN (R 4.0.2)
# rlang                  0.4.10   2020-12-30 [1] CRAN (R 4.0.2)
# rmarkdown              2.19     2022-12-15 [1] CRAN (R 4.0.4)
# rprojroot              2.0.2    2020-11-15 [1] CRAN (R 4.0.2)
# RSQLite                2.2.4    2021-03-12 [1] CRAN (R 4.0.2)
# S4Vectors            * 0.28.1   2020-12-09 [1] Bioconductor  
# scales                 1.1.1    2020-05-11 [1] CRAN (R 4.0.2)
# sessioninfo            1.1.1    2018-11-05 [1] CRAN (R 4.0.2)
# stringi                1.5.3    2020-09-09 [1] CRAN (R 4.0.2)
# stringr                1.4.0    2019-02-10 [1] CRAN (R 4.0.2)
# SummarizedExperiment * 1.20.0   2020-10-27 [1] Bioconductor  
# survival               3.2-7    2020-09-28 [1] CRAN (R 4.0.4)
# testthat               3.0.2    2021-02-14 [1] CRAN (R 4.0.2)
# tibble               * 3.1.0    2021-02-25 [1] CRAN (R 4.0.2)
# tidyr                  1.1.3    2021-03-03 [1] CRAN (R 4.0.2)
# tidyselect             1.1.0    2020-05-11 [1] CRAN (R 4.0.2)
# tweenr                 1.0.2    2021-03-23 [1] CRAN (R 4.0.4)
# usethis                2.0.1    2021-02-10 [1] CRAN (R 4.0.2)
# utf8                   1.2.1    2021-03-12 [1] CRAN (R 4.0.2)
# vctrs                  0.3.6    2020-12-17 [1] CRAN (R 4.0.2)
# vipor                  0.4.5    2017-03-22 [1] CRAN (R 4.0.2)
# withr                  2.4.1    2021-01-26 [1] CRAN (R 4.0.2)
# xfun                   0.36     2022-12-21 [1] CRAN (R 4.0.4)
# XML                    3.99-0.6 2021-03-16 [1] CRAN (R 4.0.4)
# xtable                 1.8-4    2019-04-21 [1] CRAN (R 4.0.2)
# XVector                0.30.0   2020-10-28 [1] Bioconductor  
# yaml                   2.2.1    2020-02-01 [1] CRAN (R 4.0.2)
# zip                    2.1.1    2020-08-27 [1] CRAN (R 4.0.2)
# zlibbioc               1.36.0   2020-10-28 [1] Bioconductor
```

-------------------------------------------------------------------------------------------------------------------
